pkg/cli/image/manifest/manifest.go

```diff
@@ -372,6 +381,13 @@ func ProcessManifestList(ctx context.Context, srcDigest digest.Digest, srcManife
 			}
 			klog.V(5).Infof("Including image %s for %#v from %s", manifest.Digest, manifest.Platform, ref)
 			filtered = append(filtered, manifest)
+			if PlatformSpecString(manifest.Platform) == platformSpec {
+				// If exact match, there's no regex expansion (no '*' in FilterByOS), so clear any filtered that may have
+				// been added already, since in this case we only want the exact match.
+				klog.V(5).Infof("Exact match found for %s, only including image %s from %s", platformSpec, manifest.Digest, ref)
+				filtered = []manifestlist.ManifestDescriptor{manifest}
+				break
+			}
 		}
```

Comment:
Looking at ~1st 10 lines of output with this command:
`oc image extract --filter-by-os=linux/arm quay.io/prometheus/prometheus:v2.18.1 -v=5`  
the linux/arm64, then linux/arm - that's good for cmds (like mirror) that allow ' * ' in --filter-by-os=linux/* but always want to only return the single one if an exact match was found, regardless of what's allowed via filterFn

Reply:
I don't think this method should at all be aware of filterSpec.  If you only want to match exact, the filterFn should implement exact matching.  Are there examples of manifest lists that have 2 linux/amd64 images?  In those cases the user needs to tell us which one to use, you can't just exit early.

Reply:
without this change of insisting on only a single manifest returned, when a user runs this:
`oc image extract --filter-by-os=linux/arm quay.io/prometheus/prometheus:v2.18.1 -v=5`  extract will error because it can only extract a single manifest and 2 manifests pass filterFn (arm, arm64 are available w/ that image) _note: that fails today because extract can't take a ManifestList at all, so you won't see this issue unless you're testing w/ this PR_ - you want multiple returned w/ mirror and regex expansion , but if a user passed `linux/arm` to extract & append you _only_ want that exact match - since the command will error if you get `linux/arm` + `linux/arm64` (such as w/ this prometheus image: `quay.io/prometheus/prometheus:v2.18.1`)  Here's the output of the cmd:
```console
$ mkdir ext; cd ext; oc image extract --filter-by-os=linux/arm quay.io/prometheus/prometheus:v2.18.1 -v=5
I0120 16:38:59.320392 4083752 config.go:128] looking for config.json at /home/somalley/.docker/config.json
I0120 16:38:59.320540 4083752 config.go:136] found valid config.json at /home/somalley/.docker/config.json
I0120 16:38:59.470831 4083752 credentials.go:108] Found secret to match https://quay.io/v2/auth (quay.io/auth): 
I0120 16:38:59.938937 4083752 manifest.go:405] Skipping image sha256:b4e6cd0275a26750505e539f8528e891053434ebd3972be02645bed5f02f0795 for manifestlist.PlatformSpec{Architecture:"amd64", OS:"linux", OSVersion:"", OSFeatures:[]string(nil), Variant:"", Features:[]string(nil)} from quay.io/prometheus/prometheus:v2.18.1
I0120 16:38:59.938975 4083752 manifest.go:408] Including image sha256:33cd2bbeea5c52add05be111fdccd54d289aa609f9059b22e9476865fd22b5b5 for manifestlist.PlatformSpec{Architecture:"arm64", OS:"linux", OSVersion:"", OSFeatures:[]string(nil), Variant:"", Features:[]string(nil)} from quay.io/prometheus/prometheus:v2.18.1
I0120 16:38:59.938993 4083752 manifest.go:408] Including image sha256:19bce37348f8e9fc652a0a562cad12a85abd3bde2c5ed905b36df24efec30b48 for manifestlist.PlatformSpec{Architecture:"arm", OS:"linux", OSVersion:"", OSFeatures:[]string(nil), Variant:"", Features:[]string(nil)} from quay.io/prometheus/prometheus:v2.18.1
I0120 16:38:59.939008 4083752 manifest.go:413] Exact match found for linux/arm, only including image sha256:19bce37348f8e9fc652a0a562cad12a85abd3bde2c5ed905b36df24efec30b48 from quay.io/prometheus/prometheus:v2.18.1
I0120 16:38:59.939069 4083752 manifest.go:439] Filtered manifest list to new digest sha256:f2e73eb62077979c29bc47242defbfa6875fdeee249a49bcf6067fd0ed09dead:
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 2824,
         "digest": "sha256:19bce37348f8e9fc652a0a562cad12a85abd3bde2c5ed905b36df24efec30b48",
         "platform": {
            "architecture": "arm",
            "os": "linux"
         }
      }
   ]
}
W0120 16:39:00.000542 4083752 manifest.go:456] Chose linux/arm manifest from the manifest list.
```
You can see that 2 possilble manifests were included, and we want that if the user passed --filter-by-os=linux/ar* to a mirror cmd, but if user passed `--filter-by-os=linux/arm` only the exact match should be chosen (otherwise extract &append fail, bc they only can handle a single manifest)

Reply:
> the user needs to tell us which one to use, you can't just exit early.

Right, so if the user _does_ tell us which one to use (an exact match is found) then the command should not fail bc it found both `linux/arm` and `linux/arm64`

Reply:
@sallyom we clearly state that `--filter-by-os` is a regexp, if we pass `linux/arm` and we match arm and arm64 then the bug is in the filter. If you're not passing `linux/arm.*` it should be equivalent with `^linux/arm$`, is that right @smarterclayton

Reply:
I'd like to note that it's perfectly valid to have two images for `linux/arm`:

```console
$ boater get-manifest -a nginx | jq .manifests[].platform -c
{"architecture":"amd64","os":"linux"}
{"architecture":"arm","os":"linux","variant":"v5"}
{"architecture":"arm","os":"linux","variant":"v7"}
{"architecture":"arm64","os":"linux","variant":"v8"}
{"architecture":"386","os":"linux"}
{"architecture":"mips64le","os":"linux"}
{"architecture":"ppc64le","os":"linux"}
{"architecture":"s390x","os":"linux"}
```

or for `windows/amd64`:

```console
$ boater get-manifest -a mongo | jq .manifests[].platform -c
{"architecture":"amd64","os":"linux"}
{"architecture":"arm64","os":"linux","variant":"v8"}
{"architecture":"s390x","os":"linux"}
{"architecture":"amd64","os":"windows","os.version":"10.0.17763.1697"}
{"architecture":"amd64","os":"windows","os.version":"10.0.14393.4169"}
```

Reply:
yes, I meant 2 digests for same platform, the duplicate linux/arm have different variants

Reply:
Here's correct output for extracting multi-arch w/ same os/arch, different variants:
```console
$ oc image extract --filter-by-os=linux/arm docker.io/library/nginx:latest --confirm
error: unable to read image docker.io/library/nginx:latest: filtered all images from manifest list and did not retrieve image that matches platform linux/arm
$ oc image extract --filter-by-os=linux/arm/v5 docker.io/library/nginx:latest --confirm
W0121 13:23:25.441694 4152421 manifest.go:465] Chose linux/arm/v5 manifest from the manifest list.
```

Reply:
I've pushed a new filter function, `IncludeSingle` that filters for exact match only, and passed to FirstManifest for extract & append @smarterclayton @soltysh ptal

Reply:
^^ oops forgot to remove that code that is no longer needed... removing/pushing now...

Reply:
> @sallyom we clearly state that --filter-by-os is a regexp, if we pass linux/arm and we match arm and arm64 then the bug is in the filter. If you're not passing linux/arm.* it should be equivalent with ^linux/arm$, is that right @smarterclayton

This is what I meant.  If you want a single image, pass a --filter-by-os that is exact.

Then oleg's point is about indicating that the point of the processing is to get a single element out, and it depends on whether the use case wants "all images matching the filter" or "single image matching the filter", but that's a multiple choice option on process that I would expect to get passed in to `ProcessManifestList`.  I.e. I want to process this manifest and either get all that match, get the first that matches, or fail if there are multiple matches and return an error so that the appropriate command can do the right thing. I would expect `FirstManifest` on `--filter-by-os=^linux/arm$` to already do the right thing (if we didn't need to fail in multiple matches, which would depend on the current behavior of the command).  In this case, in this command, `FirstManifest` with a correctly specified regex would return the first of duplicates and that's ok.

`release new --filter-by-os=X` should assume that X is an exact match, and so it should already be converting the value passed as X to `^X$` (i guess for backcompat you should probably see whether it's a regex or a constant, and if it's a constant you should be doing the transform, and if it's not a constant you should have the whole thing fail).  If we do that, do we need these other changes?

Reply:
`oc adm release new` does not expose `--filter-by-os` nor should it. The only time it needs this is when a user passes a multi-arch image to substitute (devs only do this) 
like w/ this: 
`oc adm release new --from-release registry.ci.openshift.org/ocp/release:4.7.0-0.nightly-2021-01-18-180737 --to-image localhost:5000/sub-release prometheus=quay.io/prometheus/prometheus:v2.18.1` that fails w/  `unable to parse manifestList` error

Reply:
and yes, with `oc image extract` `oc image append` you have to pass the exact match, or you'll get an error - but if you pass a regex that matches more than a single, you'll get a confusing error, see here:
**output from master**
```console
$ mkdir ext; cd ext; oc image extract quay.io/prometheus/prometheus:v2.18.1 --filter-by-os=linux/arm --confirm 
error: unable to parse image quay.io/prometheus/prometheus:v2.18.1: unknown image manifest of type *manifestlist.DeserializedManifestList from manifest sha256:5880ec936055fad18ccee798d2a63f64ed85bd28e8e0af17c6923a090b686c3d
```
the above error is wrong, it's erroring bc it found multiple that match, not because it's not handling the manifestList (linux/arm and linux/arm64)
**and output from this PR**:
```console
$ mkdir ext; cd ext; oc image extract quay.io/prometheus/prometheus:v2.18.1 --filter-by-os=linux/arm --confirm 
W0121 16:55:15.901894 4171937 manifest.go:457] Chose linux/arm manifest from the manifest list.
```

Reply:
In these instances, filter-by-os has to be considered a string, not a regex because the cmds can only handle a single and it's not enough to pass exactly what you want `linux/arm` bc that will not always yield a single w/ regex.MatchString.

Reply:
> get the first that matches, or fail if there are multiple matches and return an error
@smarterclayton I'd personally be against match first, b/c it will lead to different results each time, no matter where. I'd rather have always an explicit choice, either through expanding the filter with digest or something else.